### ISSUE TITLE
ApplyPatchAction / patch conflict: Ask user whether to place patch with git-like conflict marker into the affected file

### ISSUE THREAD

**Issue by Makman2** [coala/coala#4016]
Then he can fix it up in the IDE still, effective and convenient.

**Comment by satwikkansal** [comment]
Just curious. Currently, how does coala resolves patch conflicts across multiple bears?

**Comment by Siddhartha1234** [comment]
@Makman2 We need to check if Multiple Bears apply patch action on the same line and we should display an option to resolve conflicts whenever that happens. Since, we are just limited to one section at once.During processing, inside  `print_result` we call `autoapply_actions` which will apply patches just in the order in which Queue is processed and apply Result action on it. We can't resolve patch conflicts across sections, but we also can not assume that the patch that comes later is always the correct one and override it. What we can do it store  an extra dict with line number -> modification made at that line and update it whenever it is used by functions like `modify_line` or `replace`.If that line on which patch has been applied  by previous  results shows up again, then we can show a patch conflict displaying the previous line modification and current one and let them choose and then apply the patch according to his choice. Does that sound good?

**Comment by satwikkansal** [comment]
> If that line on which patch has been applied by previous results shows up again, then we can show a patch conflict displaying the previous line modification and current one and let them choose and then apply the patch according to his choice.

AFAIK, our `Diff` class is based on "original file coordinate system" i.e it doesn't allow modifications of lines which are already been modified (as compared to the original file). So I guess it already raises a  `ConflictError` when such cases arise (and it's agnostic of the section scope as the original file still remains the same until all the sections finish execution). I think this issue is more about how the users deal with conflict errors that occur.

**Comment by nemaniarjun** [comment]
<participating>

**Comment by Makman2** [comment]
coala tries to merge the diffs by creating a new `Diff`. If that doesn't go well, a `ConflictError` is raised because of the reasons @satwikkansal mentioned.

@Siddhartha1234  I think maintaining a dict like you suggest is really hard and is in principle some specialized `Diff` class. We maybe want to catch `ConflictErrors` and store the conflicts somewhere. It would also be possible to replace a part of the `Diff` (by deleting it before, though very bad overhead because you have to care about everything a diff supports due to "original file coordinate system"), so we can let the users choose. But I'm not sure if we wanna do that (at least not now). Not sure though^^ I believe we should first place git-like markers, and then we discuss in a new issue whether we want to allow the users to choose different patches by hand (or let them merge on command line, etc etc)

**Comment by Vedarth** [comment]
I would like to work on it. I think we can achieve it by resolving the conflicts just like GitHub does. This will enable the user to edit the file and choose which patch to keep. In interactive mode we can let the user choose which patch he wants to apply just like https://github.com/coala/coala/issues/1021

**Comment by 123vivekr** [comment]
@Makman2 Please assign this issue to me

**Comment by RJ722** [comment]
@123vivekr have you already solved a difficulty/low or a difficulty/newcomer issue? Only after working with those, you should attempt this one.
